File: src/lib/__tests__/beacon.test.ts
import { describe, it, expect } from "vitest";
import {
  beaconStep,
  INITIAL_BEACON_STATE,
  rebaseBeaconState,
  trendStep,
  type BeaconState,
  type BeaconFix,
} from "../beacon";

// 목적지: 서울시청 부근. 거리는 haversine 실제값을 쓰되, 같은 경도선에서
// 위도차로 거리를 만든다(위도 0.001° ≈ 111m). dest 기준 북쪽으로 떨어진 점.
const DEST = { lat: 37.5665, lng: 126.978 };
// dest에서 정북으로 d미터 떨어진 fix 생성(경도 고정, 위도만 가감).
function fixAt(metersNorth: number, accuracy = 10): BeaconFix {
  const dLat = metersNorth / 111_320; // 위도 1° ≈ 111.32km
  return { lat: DEST.lat + dLat, lng: DEST.lng, accuracy };
}

describe("beaconStep", () => {
  it("음수 accuracy(무효 좌표 신호)는 weak이고 앵커를 잡지 않는다", () => {
    // CoreLocation은 horizontalAccuracy < 0으로 좌표 무효를 신호한다. 웹 입력엔
    // 오지 않지만 가드가 갈리면 "두 플랫폼 단일 정본"이 거짓이 된다(spec §3.2).
    const { state, announce } = beaconStep(INITIAL_BEACON_STATE, fixAt(300, -1), DEST);
    expect(announce.kind).toBe("weak");
    expect(state.anchorDistance).toBeNull();
  });

  it("첫 fix는 first·speak=true·앵커 설정", () => {
    const { state, announce } = beaconStep(INITIAL_BEACON_STATE, fixAt(300), DEST);
    expect(announce.kind).toBe("first");
    expect(announce.speak).toBe(true);
    expect(Math.round(announce.distance)).toBeGreaterThan(250);
    expect(state.anchorDistance).not.toBeNull();
  });

  it("데드밴드 내 미세 진동은 추세를 뒤집지 않는다(flapping 억제)", () => {
    let s: BeaconState = beaconStep(INITIAL_BEACON_STATE, fixAt(300), DEST).state;
    // ±10m 진동(데드밴드 15m 미만) → 전부 hold
    for (const d of [305, 295, 308, 293]) {
      const r = beaconStep(s, fixAt(d), DEST);
      expect(r.announce.kind).toBe("hold");
      s = r.state;
    }
  });

  it("앵커보다 deadBand 이상 줄면 closer", () => {
    const s = beaconStep(INITIAL_BEACON_STATE, fixAt(300), DEST).state;
    const r = beaconStep(s, fixAt(250), DEST); // 50m 감소 > 15
    expect(r.announce.kind).toBe("closer");
  });

  it("앵커보다 deadBand 이상 늘면 farther", () => {
    const s = beaconStep(INITIAL_BEACON_STATE, fixAt(300), DEST).state;
    const r = beaconStep(s, fixAt(350), DEST);
    expect(r.announce.kind).toBe("farther");
  });

  it("accuracy>100은 weak·추세 불변", () => {
    const s = beaconStep(INITIAL_BEACON_STATE, fixAt(300), DEST).state;
    const r = beaconStep(s, fixAt(200, 150), DEST);
    expect(r.announce.kind).toBe("weak");
    expect(r.state.anchorDistance).toBe(s.anchorDistance);
  });

  it("정확도가 크면 데드밴드가 커져 같은 변화도 hold", () => {
    // accuracy=60 → deadBand=60. 첫 fix도 accuracy 60으로.
    const s = beaconStep(INITIAL_BEACON_STATE, fixAt(300, 60), DEST).state;
    const r = beaconStep(s, fixAt(260, 60), DEST); // 40m 감소 < 60
    expect(r.announce.kind).toBe("hold");
  });

  it("추세 flip이면 speak=true(마일스톤 미달이어도)", () => {
    let s = beaconStep(INITIAL_BEACON_STATE, fixAt(300), DEST).state;
    s = beaconStep(s, fixAt(280), DEST).state; // closer 확립
    const r = beaconStep(s, fixAt(300), DEST); // 다시 farther = flip
    expect(r.announce.kind).toBe("farther");
    expect(r.announce.speak).toBe(true);
  });

  it("farther 지속은 50m 마일스톤 유지(경고 축은 촘촘하게)", () => {
    let s = beaconStep(INITIAL_BEACON_STATE, fixAt(300), DEST).state; // lastSpoken≈300
    let r = beaconStep(s, fixAt(330), DEST); // 30m 증가 < 50 → 침묵
    expect(r.announce.kind).toBe("farther");
    expect(r.announce.speak).toBe(false);
    s = r.state;
    r = beaconStep(s, fixAt(355), DEST); // 누적 55m ≥ 50 → speak
    expect(r.announce.speak).toBe(true);
  });

  it("closer 지속은 300m 이내에서 100m 마일스톤(50m 누적으로는 침묵)", () => {
    let s = beaconStep(INITIAL_BEACON_STATE, fixAt(300), DEST).state; // lastSpoken≈300
    let r = beaconStep(s, fixAt(280), DEST); // 20m
    expect(r.announce.kind).toBe("closer");
    expect(r.announce.speak).toBe(false);
    s = r.state;
    r = beaconStep(s, fixAt(245), DEST); // 누적 55m — 종전 50m 정책이면 발화했다
    expect(r.announce.speak).toBe(false);
    s = r.state;
    r = beaconStep(s, fixAt(195), DEST); // 누적 105m ≥ 100 → speak
    expect(r.announce.speak).toBe(true);
  });

  it("closer 지속은 1km 초과에서 500m 마일스톤", () => {
    let s = beaconStep(INITIAL_BEACON_STATE, fixAt(2000), DEST).state;
    let r = beaconStep(s, fixAt(1700), DEST); // 300m < 500 → 침묵
    expect(r.announce.kind).toBe("closer");
    expect(r.announce.speak).toBe(false);
    s = r.state;
    r = beaconStep(s, fixAt(1450), DEST); // 누적 550m ≥ 500 → speak
    expect(r.announce.speak).toBe(true);
  });

  it("closer 지속은 5km 초과에서 1km 마일스톤(차량 이동 스팸 억제)", () => {
    let s = beaconStep(INITIAL_BEACON_STATE, fixAt(8000), DEST).state;
    let r = beaconStep(s, fixAt(7200), DEST); // 800m < 1000 → 침묵
    expect(r.announce.kind).toBe("closer");
    expect(r.announce.speak).toBe(false);
    s = r.state;
    r = beaconStep(s, fixAt(6900), DEST); // 누적 1,100m ≥ 1000 → speak
    expect(r.announce.speak).toBe(true);
  });

  it("도착 임박은 nearby·정밀숫자 대신 ±accuracy·진입 시 1회 speak", () => {
    const s = beaconStep(INITIAL_BEACON_STATE, fixAt(300), DEST).state;
    const r = beaconStep(s, fixAt(15, 12), DEST); // 15m ≤ arrivalThreshold(max(20,12)=20)
    expect(r.announce.kind).toBe("nearby");
    expect(r.announce.speak).toBe(true);
    expect(r.state.nearby).toBe(true);
    // 머무는 동안 재발화 안 함
    const r2 = beaconStep(r.state, fixAt(14, 12), DEST);
    expect(r2.announce.kind).toBe("nearby");
    expect(r2.announce.speak).toBe(false);
  });

  it("nearby 이탈(threshold+deadBand 초과) 시 추세 재개", () => {
    const s0 = beaconStep(INITIAL_BEACON_STATE, fixAt(300), DEST).state;
    const near = beaconStep(s0, fixAt(15, 10), DEST).state; // nearby, anchor≈15
    const r = beaconStep(near, fixAt(120, 10), DEST); // 120 > 20+10 → 이탈, farther
    expect(r.state.nearby).toBe(false);
    expect(r.announce.kind).toBe("farther");
  });

  it("hold는 speak=false", () => {
    const s = beaconStep(INITIAL_BEACON_STATE, fixAt(300), DEST).state;
    const r = beaconStep(s, fixAt(305), DEST);
    expect(r.announce.kind).toBe("hold");
    expect(r.announce.speak).toBe(false);
  });

  it("NaN 좌표는 weak로 graceful", () => {
    const s = beaconStep(INITIAL_BEACON_STATE, fixAt(300), DEST).state;
    const r = beaconStep(s, { lat: NaN, lng: NaN, accuracy: 10 }, DEST);
    expect(r.announce.kind).toBe("weak");
    expect(r.state.anchorDistance).toBe(s.anchorDistance);
  });
});

describe("축 전환 재기준화", () => {
  it("방향만 승계하고 앵커·발화 기준을 둘 다 새 축 값으로 재설정한다", () => {
    const state: BeaconState = {
      anchorDistance: 500,
      trend: "closer",
      lastSpokenDistance: 500,
      nearby: false,
    };
    expect(rebaseBeaconState(state, 120)).toEqual({
      anchorDistance: 120,
      trend: "closer",
      lastSpokenDistance: 120,
      nearby: false,
    });
  });

  it("재기준화 직후 fix는 거짓 closer를 발화하지 않는다", () => {
    // 경로 500m가 직선 120m가 되는 전환. lastSpokenDistance를 옛 축 값으로 두면
    // 차이 380m가 즉시 마일스톤(근거리 100m)을 넘겨 거짓 음성이 나간다.
    const before: BeaconState = {
      anchorDistance: 500,
      trend: "closer",
      lastSpokenDistance: 500,
      nearby: false,
    };
    const rebased = rebaseBeaconState(before, 120);
    // 새 축에서 조금 더 가까워진 fix(데드밴드 통과).
    const out = beaconStep(rebased, { lat: 0, lng: 0, accuracy: 10 }, { lat: 0, lng: 0 });
    expect(out.announce.speak).toBe(true); // 도착 존이라 nearby 발화는 정상
    // 앵커만 재설정했다면 여기서 마일스톤이 즉시 성립한다.
    const anchorOnly: BeaconState = { ...before, anchorDistance: 120 };
    expect(anchorOnly.lastSpokenDistance).toBe(500);
    expect(rebased.lastSpokenDistance).toBe(120);
  });

  it("새 축 값을 모르면 null로 두어 다음 fix가 first 경로를 탄다", () => {
    const state: BeaconState = {
      anchorDistance: 500,
      trend: "farther",
      lastSpokenDistance: 500,
      nearby: true,
    };
    const rebased = rebaseBeaconState(state, null);
    expect(rebased.anchorDistance).toBeNull();
    expect(rebased.lastSpokenDistance).toBeNull();
    expect(rebased.trend).toBe("farther");
    expect(rebased.nearby).toBe(false);
  });
});

describe("trendStep 추출", () => {
  it("앵커가 없으면 현재 거리를 앵커로 잡고 hold", () => {
    expect(trendStep(null, "none", 100, 15)).toEqual({
      kind: "hold",
      anchor: 100,
      trend: "none",
    });
  });

  it("데드밴드를 넘어 줄면 closer이고 앵커가 전진한다", () => {
    expect(trendStep(100, "none", 84, 15)).toEqual({
      kind: "closer",
      anchor: 84,
      trend: "closer",
    });
  });

  it("데드밴드를 넘어 늘면 farther", () => {
    expect(trendStep(100, "closer", 116, 15).kind).toBe("farther");
  });

  it("데드밴드 안이면 앵커·추세가 불변이다", () => {
    expect(trendStep(100, "closer", 90, 15)).toEqual({
      kind: "hold",
      anchor: 100,
      trend: "closer",
    });
  });

  it("경계값은 포함이다", () => {
    expect(trendStep(100, "none", 85, 15).kind).toBe("closer");
    expect(trendStep(100, "none", 115, 15).kind).toBe("farther");
  });
});

File: android/kit/src/test/kotlin/space/dodoplanet/gildongmu/kit/BeaconTest.kt
package space.dodoplanet.gildongmu.kit

import kotlin.math.abs
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * 웹 `beacon.test.ts` 케이스 + iOS 고유 가드 — Kit `BeaconTests` 미러. 두 플랫폼이 같은 입력에 같은 판정을
 * 내는지가 이 스위트의 계약이다. 목적지는 서울시청 부근, 같은 경도선에서 위도차로 거리를 만든다.
 */
class BeaconTest {
    private val dest = BeaconDest(37.5665, 126.978)

    private fun fixAt(metersNorth: Double, accuracy: Double = 10.0) =
        BeaconFix(dest.lat + metersNorth / 111_320, dest.lng, accuracy)

    private fun step(state: BeaconState, metersNorth: Double, accuracy: Double = 10.0) =
        beaconStep(state, fixAt(metersNorth, accuracy), dest)

    // ── beaconStep ──

    @Test fun `첫 fix는 앵커를 잡고 발화한다`() {
        val r = step(BeaconState.initial, 300.0)
        assertEquals(AnnounceKind.first, r.announce.kind)
        assertTrue(r.announce.speak)
        assertTrue(r.announce.distance > 250)
        assertNotNull(r.state.anchorDistance)
    }

    @Test fun `데드밴드가 흔들림을 억제한다`() {
        var s = step(BeaconState.initial, 300.0).state
        for (d in listOf(305.0, 295.0, 308.0, 293.0)) {
            val r = step(s, d)
            assertEquals(AnnounceKind.hold, r.announce.kind)
            s = r.state
        }
    }

    @Test fun `데드밴드를 넘으면 closer·farther`() {
        val s = step(BeaconState.initial, 300.0).state
        assertEquals(AnnounceKind.closer, step(s, 250.0).announce.kind)
        assertEquals(AnnounceKind.farther, step(s, 350.0).announce.kind)
    }

    @Test fun `정확도 상한 초과는 weak이고 앵커를 유지한다`() {
        val s = step(BeaconState.initial, 300.0).state
        val r = step(s, 200.0, accuracy = 150.0)
        assertEquals(AnnounceKind.weak, r.announce.kind)
        assertEquals(s.anchorDistance, r.state.anchorDistance)
    }

    /** 음수 정확도는 좌표 무효 신호이고, 통과시키면 deadBand가 15로 잡혀 쓰레기 좌표가 앵커가 된다. */
    @Test fun `음수 정확도는 weak이고 앵커를 잡지 않는다`() {
        val r = step(BeaconState.initial, 300.0, accuracy = -1.0)
        assertEquals(AnnounceKind.weak, r.announce.kind)
        assertNull(r.state.anchorDistance)
    }

    @Test fun `정확도 0은 weak`() {
        assertEquals(AnnounceKind.weak, step(BeaconState.initial, 300.0, accuracy = 0.0).announce.kind)
    }

    /** 이 기능의 핵심. 정확도가 나쁘면 데드밴드가 커져 같은 변화가 추세가 아니게 된다. */
    @Test fun `데드밴드는 정확도로 스케일한다`() {
        val s = step(BeaconState.initial, 300.0, accuracy = 60.0).state
        assertEquals(AnnounceKind.hold, step(s, 260.0, accuracy = 60.0).announce.kind) // 40m 감소 < deadBand 60
    }

    @Test fun `추세 반전은 마일스톤 미달이어도 발화한다`() {
        var s = step(BeaconState.initial, 300.0).state
        s = step(s, 280.0).state
        val r = step(s, 300.0)
        assertEquals(AnnounceKind.farther, r.announce.kind)
        assertTrue(r.announce.speak)
    }

    @Test fun `farther는 50m 마일스톤을 유지한다`() {
        val s0 = step(BeaconState.initial, 300.0).state
        val r1 = step(s0, 330.0)
        assertEquals(AnnounceKind.farther, r1.announce.kind)
        assertFalse(r1.announce.speak) // 30m < 50 — 경고 축은 촘촘한 간격 유지
        assertTrue(step(r1.state, 355.0).announce.speak) // 누적 55m ≥ 50
    }

    @Test fun `closer는 근거리에서 적응형 마일스톤을 쓴다`() {
        // 300m 이내는 100m 간격 — 종전 50m 정책이면 245m에서 발화했다(웹 미러).
        val s0 = step(BeaconState.initial, 300.0).state
        val r1 = step(s0, 280.0)
        assertEquals(AnnounceKind.closer, r1.announce.kind)
        assertFalse(r1.announce.speak)
        val r2 = step(r1.state, 245.0)
        assertFalse(r2.announce.speak) // 누적 55m < 100
        assertTrue(step(r2.state, 195.0).announce.speak) // 누적 105m ≥ 100
    }

    @Test fun `closer는 장거리에서 500m 간격`() {
        val s0 = step(BeaconState.initial, 2000.0).state
        val r1 = step(s0, 1700.0)
        assertEquals(AnnounceKind.closer, r1.announce.kind)
        assertFalse(r1.announce.speak) // 누적 300m < 500
        assertTrue(step(r1.state, 1450.0).announce.speak) // 누적 550m ≥ 500
    }

    @Test fun `closer는 초장거리에서 1km 간격`() {
        val s0 = step(BeaconState.initial, 8000.0).state
        val r1 = step(s0, 7200.0)
        assertEquals(AnnounceKind.closer, r1.announce.kind)
        assertFalse(r1.announce.speak) // 누적 800m < 1000
        assertTrue(step(r1.state, 6900.0).announce.speak) // 누적 1,100m ≥ 1000
    }

    @Test fun `도착 래치는 진입 때 한 번만 발화한다`() {
        val s = step(BeaconState.initial, 300.0).state
        val r = step(s, 15.0, accuracy = 12.0)
        assertEquals(AnnounceKind.nearby, r.announce.kind)
        assertTrue(r.announce.speak)
        assertTrue(r.state.nearby)
        val r2 = step(r.state, 14.0, accuracy = 12.0)
        assertEquals(AnnounceKind.nearby, r2.announce.kind)
        assertFalse(r2.announce.speak)
    }

    @Test fun `도착 래치는 히스테리시스를 넘어야 풀린다`() {
        val s0 = step(BeaconState.initial, 300.0).state
        val near = step(s0, 15.0).state
        // threshold(20) + deadBand(15) 이내는 아직 래치 유지 + hold 침묵
        val held = step(near, 30.0)
        assertEquals(AnnounceKind.hold, held.announce.kind)
        assertTrue(held.state.nearby)
        // 그 너머는 이탈하고 추세 재개
        val out = step(near, 120.0)
        assertFalse(out.state.nearby)
        assertEquals(AnnounceKind.farther, out.announce.kind)
    }

    /** 도착 임계값도 accuracy로 스케일한다. 정확도가 나쁘면 더 멀리서 "도착"이다. */
    @Test fun `도착 임계는 정확도로 스케일한다`() {
        val s = step(BeaconState.initial, 300.0).state
        assertEquals(AnnounceKind.nearby, step(s, 40.0, accuracy = 60.0).announce.kind)
        assertEquals(AnnounceKind.closer, step(s, 40.0, accuracy = 5.0).announce.kind)
    }

    @Test fun `hold는 발화하지 않는다`() {
        val s = step(BeaconState.initial, 300.0).state
        val r = step(s, 305.0)
        assertEquals(AnnounceKind.hold, r.announce.kind)
        assertFalse(r.announce.speak)
    }

    @Test fun `비유한 좌표는 weak`() {
        val s = step(BeaconState.initial, 300.0).state
        val r = beaconStep(s, BeaconFix(Double.NaN, Double.NaN, 10.0), dest)
        assertEquals(AnnounceKind.weak, r.announce.kind)
        assertEquals(s.anchorDistance, r.state.anchorDistance)
    }

    // ── 하버사인: 기준값은 웹 `haversineMeters` 실행 출력, 허용 오차 1mm ──

    @Test fun `하버사인은 웹 출력과 같다`() {
        assertEquals(0.0, haversineMeters(37.5, 127.0, 37.5, 127.0))
        assertTrue(abs(haversineMeters(37.5, 127.0, 37.501, 127.0) - 111.19492664429958) < 0.001)
        assertTrue(abs(haversineMeters(37.5665, 126.978, 37.4979, 127.0276) - 8792.890880750394) < 0.001)
    }

    // ── fix 신선도 ──

    @Test fun `신선하고 정확한 fix는 쓸 수 있다`() {
        assertTrue(isUsableFix(10.0, 0.0))
        assertTrue(isUsableFix(65.0, 4.9))
    }

    @Test fun `낡은 fix는 거부한다`() {
        assertFalse(isUsableFix(10.0, 5.1))
        assertFalse(isUsableFix(10.0, 120.0))
    }

    @Test fun `미래 timestamp는 거부한다`() {
        assertFalse(isUsableFix(10.0, -30.0))
    }

    @Test fun `무효 정확도는 거부한다`() {
        assertFalse(isUsableFix(-1.0, 0.0))
        assertFalse(isUsableFix(0.0, 0.0))
    }

    @Test fun `창은 설정 가능하다`() {
        assertTrue(isUsableFix(10.0, 20.0, maxAge = 30.0))
    }

    // ── 추세 판정 추출(두 모드가 공유하는 유일한 축) ──

    @Test fun `앵커가 없으면 현재 거리를 앵커로 잡고 hold`() {
        assertEquals(TrendStepResult(TrendKind.hold, 100.0, BeaconTrend.none), trendStep(null, BeaconTrend.none, 100.0, 15.0))
    }

    @Test fun `데드밴드를 넘어 줄면 closer이고 앵커가 전진한다`() {
        assertEquals(TrendStepResult(TrendKind.closer, 84.0, BeaconTrend.closer), trendStep(100.0, BeaconTrend.none, 84.0, 15.0))
    }

    @Test fun `데드밴드를 넘어 늘면 farther`() {
        val r = trendStep(100.0, BeaconTrend.closer, 116.0, 15.0)
        assertEquals(TrendKind.farther, r.kind)
        assertEquals(BeaconTrend.farther, r.trend)
    }

    @Test fun `데드밴드 안이면 hold이고 앵커·추세가 불변이다`() {
        assertEquals(TrendStepResult(TrendKind.hold, 100.0, BeaconTrend.closer), trendStep(100.0, BeaconTrend.closer, 90.0, 15.0))
    }

    @Test fun `경계값은 포함이다`() {
        assertEquals(TrendKind.closer, trendStep(100.0, BeaconTrend.none, 85.0, 15.0).kind)
        assertEquals(TrendKind.farther, trendStep(100.0, BeaconTrend.none, 115.0, 15.0).kind)
    }

    // ── 축 전환 재기준화 ──

    @Test fun `방향만 승계하고 앵커·발화 기준을 둘 다 새 축 값으로 재설정한다`() {
        val state = BeaconState.initial.copy(anchorDistance = 500.0, trend = BeaconTrend.closer, lastSpokenDistance = 500.0)
        val out = rebaseBeaconState(state, 120.0)
        assertEquals(120.0, out.anchorDistance)
        assertEquals(120.0, out.lastSpokenDistance)
        assertEquals(BeaconTrend.closer, out.trend)
        assertFalse(out.nearby)
    }

    @Test fun `새 축 값을 모르면 null로 두어 다음 fix가 first 경로를 탄다`() {
        val state = BeaconState(anchorDistance = 500.0, trend = BeaconTrend.farther, lastSpokenDistance = 500.0, nearby = true)
        val out = rebaseBeaconState(state, null)
        assertNull(out.anchorDistance)
        assertNull(out.lastSpokenDistance)
        assertEquals(BeaconTrend.farther, out.trend)
        assertFalse(out.nearby)
    }
}
